### PR TITLE
Add crate that enables the full float APIs in `#[no_std]`

### PR DESCRIPTION
**Pull request by robertbastian**
It provides an extension trait for `f32` and `f64` that replicates the whole std API, backed by `libm`. This allows us to remove explicit `libm` calls from our business logic, and use the std implementations on std targets via shadowing.

This PR uses `core_float` in `icu_segmenter`, I will also apply it in `calendrical_calculations` after #3985 is merged.

### CODE REVIEW

File: utils/core_maths/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,1142 @@
...
+        q
+    }
+
+    fn rem_euclid(self, rhs: Self) -> Self {
+        let r = self % rhs;
+        if r < 0.0 {
+            r + rhs.abs()
+        } else {
+            r
+        }
```
Comment (sffc) [inline utils/core_maths/src/lib.rs:755]: General issue: We should minimize the amount of math code in this crate. If we must add any, it should be well tested. One way to test it is to run it against the standard library implementation on a suitably diverse set of inputs.

Specific issue: Pretty sure the div/rem euclid impls here aren't right. The [standard library impl Rem](https://doc.rust-lang.org/std/primitive.f64.html#impl-Rem%3Cf64%3E-for-f64) says:

> The remainder has the same sign as the dividend and is computed as: `x - (x / y).trunc() * y`

But rem_euclid is a different thing. It is the remainder relative to the floor, not the trunc. You can't derive that by taking `.abs()`.

File: utils/core_maths/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,1142 @@
...
+    }
+
+    #[inline]
+    fn atanh(self) -> Self {
+        0.5 * ((2.0 * self) / (1.0 - self)).ln_1p()
+    }
+}
+
+impl CoreFloat for f64 {
```
Comment (sffc) [inline utils/core_maths/src/lib.rs:915]: Issue: Shouldn't you be gating this impl on `feature = "std"` or something?

**Review by sffc — CHANGES_REQUESTED** [review]

File: utils/core_maths/src/lib.rs (hunk 2 above)
Comment (robertbastian) [reply utils/core_maths/src/lib.rs:915]: what would that achieve? the idea of this whole crate is to make these methods available without std.

File: utils/core_maths/src/lib.rs (hunk 1 above)
Comment (robertbastian) [reply utils/core_maths/src/lib.rs:755]: All code here is copied from `std`, with calls to intrinsics replaced by libm. You're linking to `Rem`, which is a different operation to [`rem_euclid`](https://doc.rust-lang.org/src/std/f64.rs.html#328).

File: utils/core_maths/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,1142 @@
...
+        let r = self % rhs;
+        if r < 0.0 {
+            r + rhs.abs()
+        } else {
+            r
+        }
+    }
+
+    #[inline]
+    fn powi(self, mut exp: i32) -> Self {
```
Comment (robertbastian) [inline utils/core_maths/src/lib.rs:759]: this is the only code not copied from `std::f64`, libm doesn't have a `powi` so this is from `u64::pow`.

File: utils/core_maths/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+authors.workspace = true
+edition.workspace = true
+repository.workspace = true
+homepage.workspace = true
+license-file.workspace = true
+categories.workspace = true
+include.workspace = true
+
+[dependencies]
+libm = "0.2"
```
Comment (Manishearth) [inline utils/core_maths/Cargo.toml:20]: should this dep be `cfg(target_os = none)`?

File: Cargo.toml [hunk 5]
```diff
@@ -199,6 +200,7 @@ icu_provider_macros = { version = "~1.2.0", path = "provider/macros", default-fe
 
 # Utils
 calendrical_calculations = { version = "0.1.0", path = "utils/calendrical_calculations", default-features = false }
+core_maths = { version = "0.0.0", path = "utils/core_maths", default-features = false }
```
[the remark this answers is not in the document]
Comment (robertbastian) [reply Cargo.toml:203]: I'd prefer to call this `core_float`, but that name is already taken by someone rebuilding `num_traits` or something. Any ideas?

File: utils/core_maths/Cargo.toml (hunk 4 above)
Comment (robertbastian) [reply utils/core_maths/Cargo.toml:20]: Sure

File: Cargo.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (Manishearth) [reply Cargo.toml]: `core_math` seems fine imo

`core_float_methods` if you really want

[the remark this answers is not in the document]
Comment (robertbastian) [reply Cargo.toml]: `libm_core`?

File: utils/core_maths/src/lib.rs (hunk 1 above)
Comment (sffc) [reply utils/core_maths/src/lib.rs]: Ok `r + rhs.abs()` I think is correct. We care a lot about rem_euclid for calendrical calculations.

Comment (robertbastian) [reply utils/core_maths/src/lib.rs]: `rhs.abs()` is the more general form, we use `rhs` in our version because we require `rhs > 0`

**Comment by sffc** [comment]
I'm not convinced that this PR is a positive change for these reasons:

1. Currently each component specifies and tests its own set of math operations relevant for that component. I think this status quo is good, because we only add these functions when they are needed, so all the code is covered and tested. With a mega trait[^1] like this, most of this code is not going to be covered or used.
2. Related to the above point, this code is not covered by tests. We want high test coverage both to satisfy Coveralls but also because it helps ensure that we are consistent with `std`. Furthermore, there isn't anything specific to ensure we keep this in sync with rust std.
3. We are in the business of internationalization. I think we need to be conservative with non-i18n code landing in our repo. Existing utilities like Yoke, ZeroVec, ZeroFrom, TinyStr, and so forth are widely used by most all ICU4X components, and ones like FixedDecimal, ZeroTrie, and Writeable are definitely i18n-adjacent. `core_maths` is neither widely used (it is needed for only segmenter and calendrical calculations) nor i18n-adjacent (in both cases it is for implementing a non-i18n algorithm, ML and astronomy). Anywhere we need non-i18n utilities (like matrix math), we implement a narrow subset in private modules.

[^1]: Also, mega traits are slightly dangerous because if you ever take them as a trait object, you get all the code for all their functions with no code slicing available. This could be mitigated by making the mega trait not be object safe.

**Comment by robertbastian** [comment]
> Furthermore, there isn't anything specific to ensure we keep this in sync with rust std.

We don't have to keep it in sync, Rust `std` is not going to change behaviour overnight.

We have other utility crates that are not used by every component, like crlify, deduplicating_array, even litemap is only used by two components.

This trait is not object safe because it takes `self` by value.

File: utils/core_maths/Cargo.toml (hunk 4 above)
Comment (robertbastian) [reply utils/core_maths/Cargo.toml]: actually this doesn't work, it's possible to build `no_std` on a target with an OS.

**Comment by sffc** [comment]
> We have other utility crates that are not used by every component, like crlify, deduplicating_array, even litemap is only used by two components.

`crlify` is less than 50 lines of code, so the risks and maintenance burden are much less.

`deduplicating_array` is a bit larger, yes. I think the reason it ended up there was that we needed it as public API because it is in our data struct definition and couldn't figure out a better place to put it. But, we don't need `core_maths` in our public API.

`litemap` had previously been a utility used widely, until ZeroMap came along and replaced most of our usages. We have a path toward unifying our map solution in #3128.

**Comment by sffc** [comment]
> We don't have to keep it in sync, Rust `std` is not going to change behaviour overnight.

The most likely situation I'm thinking of is Rust std adding new mathematical operations. Not super common but it does seem to happen from time to time (f64::clamp, f64::total_cmp, f64::from_ne_bytes and friends, f64::to_int_unchecked, f64::EPSILON, f64::copysign, f64::div_euclid, and f64::rem_euclid are all examples of APIs added after 1.0), and I don't want the ICU4X team to take on the responsibility to check for that.

**Comment by robertbastian** [comment]
I think you might be overestimating the maintenance burden. I'm happy to own this crate.

**Comment by sffc** [comment]
I would be happy with any of the following options (not in any particular order):

1. Write tests for this crate that compare your function output with std function output. I definitely do not want to land any more math code with spotty test coverage given https://github.com/unicode-org/icu4x/pull/3644. If you add tests and @Manishearth is aligned on the crate concept, we can land this.
2. Make a new repository like `robertbastian/coremaths` (going through the releasing process) and then we take it in as a third-party dependency.
3. Close this PR and optionally open a new one polishing up the private component-specific traits.

(however, there are other 1.3 blocking issues that are all higher priority than landing this PR, in my opinion)

**Comment by robertbastian** [comment]
I will not write tests for std against std, that just does not make sense. It's also not possible to exhaustively test `f32`x`f32` or `f64` methods.

This is very simple code, which can easily reviewed by looking at the mathematical definition that each method represents. It is not comparable to the complicated business logic in #3644.

> Make a new repository like robertbastian/coremaths (going through the releasing process) and then we take it in as a third-party dependency.

I'm happy to do that but what exactly does that achieve? I'll still want to depend on this from ICU4X.

**Pull request closed without merging**